activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb

```diff
@@ -985,13 +985,30 @@ def foreign_key_options(from_table, to_table, options) # :nodoc:
       end
 
       def dump_schema_information #:nodoc:
+        versions = ActiveRecord::SchemaMigration.order('version').pluck(:version)
+        if supports_multi_insert?
+          multi_value_version_insert(versions)
+        else
+          single_value_version_insert(versions)
+        end
+      end
+
+      def multi_value_version_insert(versions)
         sm_table = ActiveRecord::Migrator.schema_migrations_table_name
 
         sql = "INSERT INTO #{sm_table} (version) VALUES "
-        sql << ActiveRecord::SchemaMigration.order('version').pluck(:version).map {|v| "('#{v}')" }.join(', ')
+        sql << versions.map {|v| "('#{v}')" }.join(', ')
         sql << ";\n\n"
       end
 
+      def single_value_version_insert(versions)
+        sm_table = ActiveRecord::Migrator.schema_migrations_table_name
+
+        versions.map { |version|
+          "INSERT INTO #{sm_table} (version) VALUES ('#{version}');"
+        }.join "\n\n"
+      end
+
```

Comment:
Splitting the insert methods means the caller is responsible for figuring out which to use. The caller is concerned with inserting versions, not with which insertion technique to use. Let's combine a single `insert_versions(versions)` for them to call:

```
def dump_schema_information
  insert_versions_sql ActiveRecord::SchemaMigration.order('version').pluck(:version)
end

def insert_versions_sql(versions)
  sm_table = ActiveRecord::Migrator.schema_migrations_table_name

  if supports_multi_insert?
    …
  else
    …
  end
end

def assume_migrated_upto_version(version, migrations_paths)
  …
  execute insert_versions_sql(versions)
end
```